### PR TITLE
Fix for issue #39827

### PR DESCRIPTION
**Pull request by taleks**
*Cause of the issue*
    
While preparing for `trans_intrinsic_call()` invoke arguments are processed with `trans_argument()` method which excludes zero-sized types from argument list (to be more correct - all arguments for which `ArgKind` is `Ignore` are filtered out). As result `volatile_store()` intrinsic gets one argument instead of expected address and value.
    
*How it is fixed*
 
Modification of the `trans_argument()` method may cause side effects, therefore change was implemented in `volatile_store()` intrinsic building code itself. Now it checks function signature and if it was specialised with zero-sized type, then emits `C_nil()` instead of accessing non-existing second argument.

### CODE REVIEW

**Comment by arielb1** [comment]
Nice PR!

I'm not sure we want the "suspicious monomorphization" warning - it can spam the user in dead generic code. r=me without it.

File: src/libcore/ptr.rs [hunk 1]
```diff
@@ -384,6 +384,12 @@ pub unsafe fn write_unaligned<T>(dst: *mut T, src: T) {
 /// over time. That being said, the semantics will almost always end up pretty
 /// similar to [C11's definition of volatile][c11].
 ///
+/// Compiler shouldn't change relative order or number of volatile memory
```
Comment (arielb1) [inline src/libcore/ptr.rs]: eng:

The compiler shouldn't change the relative order or number of volatile memory operations. However, volatile memory operations on zero-sized types (e.g. if a zero-sized type is passed to `read_volatile`) are no-ops and may be ignored.

**Comment by taleks** [comment]
Thanks for comments.
I've removed warnings related code from `librustc_trans/intrinsic.rs` and updated changes in documentation  for `write_volatile()`/`read_volatile()`.

**Pull request merged**